shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/Spark311Shims.scala

```diff
@@ -204,29 +204,45 @@ class Spark311Shims extends Spark301Shims {
     // Spark 3.1.1-specific LEAD expression, using custom OffsetWindowFunctionMeta.
     GpuOverrides.expr[Lead](
       "Window function that returns N entries ahead of this one",
-      ExprChecks.windowOnly((TypeSig.numeric + TypeSig.BOOLEAN +
-        TypeSig.DATE + TypeSig.TIMESTAMP + TypeSig.ARRAY).nested(), TypeSig.all,
-        Seq(ParamCheck("input", (TypeSig.numeric + TypeSig.BOOLEAN +
-          TypeSig.DATE + TypeSig.TIMESTAMP + TypeSig.ARRAY).nested(), TypeSig.all),
+      ExprChecks.windowOnly(
+        (TypeSig.commonCudfTypes + TypeSig.DECIMAL + TypeSig.NULL +
+          TypeSig.ARRAY + TypeSig.STRUCT).nested(),
```

Comment:
This is adding in structs as well as string, but I didn't see any tests for structs, or arrays of structs or structs with arrays in them/etc.

Reply:
In [window_function_test.py](https://github.com/NVIDIA/spark-rapids/pull/2842/files#diff-1d2ae10fd5085346e0f2055455ac2f881d950ed8215d61e3cb708917ed1d41bcR128), `lead_lag_data_gens` was modified thus:
```python
lead_lag_data_gens = [long_gen, DoubleGen(no_nans=True, special_cases=[]),
        boolean_gen, timestamp_gen, string_gen, DecimalGen(precision=18, scale=3),
        StructGen(children=[
            ['child_int', IntegerGen()],
            ['child_time', DateGen()],
            ['child_string', StringGen()]
        ])]
```
Also, `lead_lag_array_data_gens` generates `ARRAY` versions of the data, from `lead_lag_data_gens`. So the `ARRAY<STRUCT>` test is covered when `test_window_aggs_for_rows_lead_lag_on_arrays` runs.

That said, I should add an `ARRAY` type to the `StructGen` in `lead_lag_data_gens`.

Reply:
As an aside, it appears that the `Struct` tests seem to be failing at the moment, while they weren't last night. I'll add the `Struct<Array>` test as soon as that's sorted out.

Reply:
Hmm... This is interesting. It's failing because one of the tests is to specify a default value for `LAG()`. I can't seem to figure out how to specify a `STRUCT` literal in SQL. The following doesn't work:
`LAG(STRUCT(dollars, dateStr), 3, STRUCT(CAST(1 AS INT),'') )`:
```
org.apache.spark.sql.AnalysisException: cannot resolve 'lag(struct(mytable.`dollars`, mytable.`dateStr`), 3, struct(CAST(1 AS INT), ''))' due to data type mismatch: argument 3 requires (struct<dollars:int,dateStr:string> or null) type, however, 'struct(CAST(1 AS INT), '')' is of struct<col1:int,col2:string> type.; line 1 pos 22;
```

Reply:
I am inclined to use `None` for `STRUCT` column `LAG()` defaults, because `[1, "two"]` isn't a valid `STRUCT` literal.

Another interesting problem to sort out is why `SortOrder` does not support ordering on `STRUCT` columns in windowing contexts:
```
!Expression <WindowSpecDefinition> windowspecdefinition(uid#81L, dateLong#83L ASC NULLS FIRST, _w2#95 ASC NULLS FIRST, specifiedwindowframe(RowFrame, -3, -3)) cannot run on GPU because expression SortOrder _w2#95 ASC NULLS FIRST produces an unsupported type StructType(StructField(dollars,IntegerType,true), StructField(dateStr,StringType,true))
```

Reply:
@mythrocks I think it's because `WindowSpecDefinition` https://github.com/NVIDIA/spark-rapids/blob/b3ae804b76c2b6c2fa396c17d7a7be1a8a32b1c3/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala#L874-L878
is configured to deal with nested outputs https://github.com/NVIDIA/spark-rapids/blob/b3ae804b76c2b6c2fa396c17d7a7be1a8a32b1c3/sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala#L803-L805

Reply:
I'm past that now. 
```scala
object WindowSpecCheck extends ExprChecks {
  val check: TypeSig =
    TypeSig.commonCudfTypes + TypeSig.DECIMAL + TypeSig.NULL +
      TypeSig.STRUCT.nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL + TypeSig.NULL)
```
The new problem I'm wrestling with is whether to allow a `STRUCT` order-by column for `RANGE` queries. For instance, the window function tests currently have the equivalent of the following:
```sql
SELECT MIN(my_struct) OVER (PARTITION BY my_id ORDER BY my struct) ...
```
This translates to a `RANGE` query `(UNBOUNDED, CURRENT ROW)`.
Technically, sorting on a `STRUCT` column can be GPU accelerated, IFF the column isn't further nested. I'm inclined to enable this sorting in window functions, for this narrow case.

I'll update here when I have this going.

Reply:
P.S. Thanks for reviewing/investigating, @gerashegalov.

Reply:
I'm bailing on this. The scope is ballooning, and there would need to be changes in the `cudf::rolling_window()` to support this for structs. I'll redo the test to skip `min()`/`max()` operations for structs.